asyncapi.yaml

```
@@ -0,0 +1,93 @@
+asyncapi: '2.0.0' # Go parser does not support 2.1.0 at this moment.
+info:
+  title: AsyncAPI Event-Gateway demo API
+  version: 1.0.0-alpha
+  description: This API lets users interact with an instance of the [AsyncAPI Event-Gateway](https://github.com/asyncapi/event-gateway).
+defaultContentType: application/json
+servers:
+  asyncapi-event-gateway-demo:
+    url: 'event-gateway-demo.asyncapi.org:28002'
+    protocol: kafka
+    description: Asyncapi [Event-Gateway](https://github.com/asyncapi/event-gateway) demo Kafka proxy. Expects messages based on StreetLights tutorial available at https://www.asyncapi.com/docs/tutorials/streetlights.
```

```suggestion
    description: AsyncAPI [Event-Gateway](https://github.com/asyncapi/event-gateway) demo Kafka proxy. Expects messages based on StreetLights tutorial available at https://www.asyncapi.com/docs/tutorials/streetlights.
```

AsyncAPI is spelled wrong in other descriptions. 

For streetlight maybe link here http://bit.ly/asyncapi ?

---

Also I thought on referencing the lightMeasured messages by using `$ref: 'https://raw.githubusercontent.com/asyncapi/asyncapi/2.0.0/examples/2.0.0/streetlights.yml#/components/messages/lightMeasured'`, however the go parser complains with:

```
#/components/messageTraits/commonHeaders: key not found\n#/components/schemas/lightMeasuredPayload: key not found
```

Isn't possible to do that? it's just the go parser?

---

@smoya Yes, referencing the part from file is supported (in JS it is, I don't know which library for dereferencing is used in GO Parser), but as I see in the error context, that `#/components/messageTraits/commonHeaders` complains about missing `#/components/schemas/lightMeasuredPayload` schema, in your AsyncAPI document, not from external file - or maybe this error is from file context

---

Ok I see, in external file you have this definition of message:

```yaml
    lightMeasured:
      name: lightMeasured
      title: Light measured
      summary: Inform about environmental lighting conditions of a particular streetlight.
      contentType: application/json
      traits:
        - $ref: '#/components/messageTraits/commonHeaders'
      payload:
        $ref: "#/components/schemas/lightMeasuredPayload"
```

so dereferencer tool tries to dereference payload and traits from context of your file, not from context of external file.

---

> I thought using references will make other references (references made by the referenced object) work as well. If this assumption is wrong, that means I will need to create those schemas in my file and reference them to the remote file.

You can try to parse your spec by ParserJS and check if it can dereference in proper way - if yes, then Go Parser uses "broken" tool for dereferencing.

---

I just did and it seems it works great on parser-js:

```js
// ...
// console.log(doc.channels()['event-gateway-demo'].publish().message());
```

```js
Message {
  _json: {
    name: 'lightMeasured',
    title: 'Light measured',
    summary: 'Inform about environmental lighting conditions of a particular streetlight.',
    contentType: 'application/json',
    payload: {
      type: 'object',
      properties: [Object],
      'x-parser-schema-id': '<anonymous-schema-3>'
    },
    headers: {
      type: 'object',
      properties: [Object],
      'x-parser-schema-id': '<anonymous-schema-1>'
    },
    'x-parser-original-traits': [ [Object] ],
    'x-parser-original-schema-format': 'application/vnd.aai.asyncapi;version=2.0.0',
    'x-parser-original-payload': { type: 'object', properties: [Object] },
    schemaFormat: 'application/vnd.aai.asyncapi;version=2.0.0',
    'x-parser-message-parsed': true
  }
}
```

I will open an issue on the Go parser. Meanwhile I will just copy and paste everything I need from streetlights yaml. 🤷

---

For reference: https://github.com/asyncapi/parser-go/issues/82

---

I updated the doc using all references I could considering the bug in the parser-go pkg.